compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13483,7 +13483,7 @@ OMR::Z::TreeEvaluator::forwardArrayCopySequenceGenerator(TR::Node *node, TR::Cod
       // 2. byteLength > 256 && byteLength is not multiple of 256
       uint8_t residueLength = static_cast<uint8_t>(byteLen & 0xFF);
       if (byteLen <= 256 || residueLength != 0)
-         generateSS1Instruction(cg, TR::InstOpCode::MVC, node, residueLength-1, 
+         generateSS1Instruction(cg, TR::InstOpCode::MVC, node, static_cast<uint8_t>(residueLength-1), 
```

Comment:
@fjeremic `generateSS1Instruction` expects 16 bit unsigned integer argument. When we have constant byte length without explicit casting it sends `65535`  instead of `255` to generateSS1Instruction. Just explicitly cast value here without speculating binary encoding generate current value for us.

Reply:
Not sure I'm following here. Why does explicit casting here change anything? `residueLength` is already a `uint8_t`, and hence so is `residueLength - 1` so the cast doesn't change the answer at all I would think.

Reply:
Looking at the function declaration 
https://github.com/eclipse/omr/blob/c2ec9f716193f5c995ee22619ebedf27bdab1d48/compiler/z/codegen/S390GenerateInstructions.hpp#L647
 In case when `residueLength-1` will be casted to uint16_t. So is `byteLen` is multiple of `256` means `residueLength` will be `0`. In this case here our intended argument is `255` but it will send `65535` instead. 
Verified this in the instruction generation time : 
```
[0x787909c0]   MVC     #638 0(65536,GPR_0044),#639 0(GPR_0042)
```

However it does generate correct binary encoding but looking at the Instruction Generation it is misleading. I took a look at `SS-a` type instructions. I do not see any one uses 16 bit length. Is it possible to clean up SS1 instruction generator to reflect this?

Reply:
I see. Thanks for the explanation. We should avoid these headaches and modify the `generateSS1Instruction` API to take a `uint8_t` since the `L` field of the SS encoding is 8 bits wide and unsigned. Not sure why we even give the option to the user to shoot themselves in the foot with the `uint16_t` parameter type. Can you open up an item for this in OMR and on our ZenHub?